### PR TITLE
feat: add GroupKFold cross-validation + AbstractCVStrategy

### PR DESCRIPTION
**Pull request by biocompt**
First of four CV PRs (per #17). Adds the cross-validation contract and `GroupKFold` as the foundation strategy. Architectural decisions are spelled out in https://github.com/davide-grheco/DataSplits.jl/issues/17#issuecomment-4370234967.

### 1. `AbstractCVStrategy <: AbstractSplitStrategy`

Marker subtype that selects the CV-specific `partition` method:

```julia
partition(data, alg::AbstractCVStrategy;
          target=nothing, time=nothing, groups=nothing,
          rng=Random.default_rng()) -> CrossValidationSplit
```

This method does **not** accept `train` / `test` / `validation` keywords — fold sizes are fixed by the strategy itself (typically through `k`). The 2-cohort and 3-cohort `partition` methods are unchanged. Slot resolution (`_resolve_slot` / `_slot_for` / `_to_feature_matrix`) is reused as-is; only `_resolve_sizes` is bypassed (it doesn't apply).

### 2. `GroupKFold(k)`

Group-aware k-fold CV. Whole groups go to a single fold; deterministic balancing (sklearn-style: sort groups by descending size, place each in the smallest fold so far). Validates `length(groups) == numobs(data)` locally — anticipates #22.

### 3. `Base.getindex` / `firstindex` / `lastindex` / `eltype` on `CrossValidationSplit`

Closes the gap from your RFC comment: `cvs[1]`, `cvs[end]`, `cvs[range]` (slicing returns a `CrossValidationSplit` of the selected folds), `first(cvs)`, `last(cvs)`, `collect(cvs)` all work. Iteration was already present.

## Tests

- 36 new tests for `GroupKFold`: contract, group respect, balancing, fallback, splitview/splitdata, parameter validation, mismatched-length input, rejection of `train`/`test` kwargs.
- 8 new tests for `CrossValidationSplit` indexing.
- Existing suite (~350 tests) intact, no regressions.

## What's NOT in this PR

- `LeaveOneGroupOut`, `TimeSeriesCV`, `StratifiedKFold` — ready in local branches, will follow as PRs 2–4 once this lands (per the "4 small PRs" preference in #17).
- Docs pages — held back until #19 unblocks the `docs/make.jl` build.

## Verification

```bash
julia --project=. -e 'using Pkg; Pkg.test()'
# 510+ tests, all passing.
```

Closes part of #17.

### CODE REVIEW

File: src/strategies/GroupKFold.jl [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+# Fields
+- `k::Int`: Number of folds (must be ≥ 2 and ≤ number of unique groups).
+
+# Notes
+- Fold assignment is deterministic: groups are sorted by size (descending)
+  and each is placed in the currently smallest fold. This yields folds of
+  roughly equal observation count, like sklearn's `GroupKFold`.
+- The `rng` keyword is accepted for interface uniformity but unused; results
+  are reproducible without it.
```
Comment (davide-grheco) [inline src/strategies/GroupKFold.jl]: I would expect running KFold different times would give me different results. I'm not sure I fully like this. We already have GroupShuffleSplit, why don't we use that and if we have to improve logic we improve it there rather than duplicating it?

File: src/core.jl [hunk 2]
```diff
@@ -571,3 +599,59 @@ function partition(
...
+  alg::AbstractCVStrategy;
+  target = nothing,
+  time = nothing,
+  groups = nothing,
+  rng = Random.default_rng(),
+)
+  isempty(data) &&
+    throw(SplitInputError("Data must not be empty. Please provide a non-empty dataset."))
+  N = numobs(data)
+  N >= 2 || throw(SplitInputError("Cannot split fewer than 2 observations."))
```
Comment (davide-grheco) [inline src/core.jl]: At this point, we should extract this into a helper validation function.

**Review by davide-grheco — COMMENTED** [review]
It's mostly good, but I'm not sure that the grouping should be deterministic.
I understand that makes it simpler to split them, but there may be other ways.

scikit-learn has a shuffle parameter to GroupKFold: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GroupKFold.html
We could just follow their implementation and use the passed rng to seed the shuffling.

File: src/core.jl (hunk 2 above)
Comment (biocompt) [reply src/core.jl]: Done in 3586a89. Pulled the empty + N≥2 check into `_assert_partitionable(data) -> N` in `validation.jl` and replaced all three sites. Same behaviour and error messages.

File: src/strategies/GroupKFold.jl (hunk 1 above)
Comment (biocompt) [reply src/strategies/GroupKFold.jl]: Good push — added `shuffle::Bool = false` in c906674. Off by default so results stay reproducible without an `rng`, but `shuffle = true` taps the rng you're already threading through `partition`. Same shape as sklearn's `GroupKFold(shuffle=, random_state=)`.

On the `GroupShuffleSplit` consolidation though, I'd push back a bit. They feel like separate operations to me rather than a duplication:

- `GroupShuffleSplit` returns one train/test split, and some groups may never end up in test. Running it `k` times gives `k` independent shuffle splits with overlapping test cohorts and no coverage guarantee.
- `GroupKFold` returns `k` splits where every group lands in exactly one fold's test set — that disjoint partitioning of the test space is the whole point of k-fold.

The actually shared logic is pretty thin (just the group → indices map). The balancing loop here ("put the next group in the currently smallest fold") doesn't really fit `GroupShuffleSplit`'s "fill train until n_train" semantics. If a third group-aware strategy lands I think we'd extract a small helper at that point, but for now I'd keep them separate. Open to keep talking about it though if you see something I don't.

**Comment by biocompt** [comment]
Thanks for the review! Both points addressed in c906674 and 3586a89.

`shuffle = false` by default to preserve the no-rng-needed reproducibility, but `shuffle = true` uses the `rng` you're already passing to `partition`. The validation helper is now in `validation.jl` and shared by all three `partition` methods.

I pushed back a bit on the `GroupShuffleSplit` merge idea in the inline thread — I think they're solving different problems and consolidating would lose the test-space coverage guarantee that's the whole reason to reach for k-fold. Happy to dig in more if you disagree.
